packages/liquidator/src/liquidationStrategy.js

```diff
@@ -0,0 +1,254 @@
+const assert = require("assert");
+/**
+ * This module is responsible for sizing liquidation positions.  If a defensive strategy is enabled, it
+ * will check withdraw positions to see if they require an extension rather than full liquidation.
+ * @constructor
+ * @param {object} config - State configuration, env vars, contract vars
+ * @property {number} config.whaleDefenseFundWei - Optional value which enables strategy if above 0. Specify the
+ * reserves to hold in tokens to allow for defense strategy. 0 disables.
+ * @property {number} config.defenseActivationPercent - The % past liveness a position needs to be before its
+ * extended. Optional unless whaleDefenseFundWei is enabled.
+ * @property {number} config.liquidationDeadline - Aborts the liquidation if the transaction is mined this amount
+ * of time after this time has passed
+ * @property {number} config.liquidationLiveness - Optional unless whaleDefenseFundWei is enabled. The positions
+ * liveness duration, set by emp contract.
+ * @property {number} config.minSponsorSize - Emp minimum sponsor position size in tokens.
+ * Example:
+ * {
+ *   whaleDefenseFundWei: '10000000000000000000',
+ *   defenseActivationPercent: 80,
+ *   liquidationDeadline: 300,
+ *   liquidationLiveness: 10000,
+ *   minSponsorSize: '5000000000000000000'
+ * }
+ * @param {object} deps - Library dependencies
+ * @property {object} deps.toBN - toBN function
+ * @property {object} deps.BN - BN utilities
+ * Example:
+ * {
+ *   toBN: Web3.utils.toBN,
+ *   BN: Web3.utils.BN,
+ * }
+ * @callback emit - A way to emit logs to an external logger
+ * @param {string} severity - Log level severity
+ * @param {object} data - Any data you want to be sent in log
+ * Example:
+ * function log(severity,data){
+ *   logger[severity]({ at:'Liquidator', ...data })
+ * }
+ */
+module.exports = (
+  {
+    // how much money to set aside for defense in tokens
+    whaleDefenseFundWei = 0,
+    // when to activate defense, when withdraw is X% in time complete
+    defenseActivationPercent,
+    liquidationMinPrice = "0",
+    // `liquidationDeadline`: Aborts the liquidation if the transaction is mined this amount of time after the
+    liquidationDeadline = 300,
+    // the minimum amount of tokens to liquidate with resetting timer
+    liquidationLiveness,
+    // emp contracts min sponsor size, specified in tokens
+    minSponsorSize
+  } = {},
+  // These could probably be pulled in from web3, but the pattern is here to add
+  // Any additional library dependencies if needed
+  { toBN, BN } = {},
```

Comment:
I think that is one of the benefits, although these libraries in particular don't really make sense here. I dont have a good reason to inject them. Will probably remove in favor of importing from web3.

In general though I do like the pattern of instanciating classes as `constructor(configs,dependencies,events)`, so they are really more to stub the second parameter in that signature.

Reply:
I just realized looking back that it is this way because the parent (liquidator) gets web3 passed in and destructures it for BN and toBN. So I copied that behavior of dependency injection. Will probably keep it as is unless you have objections!

Reply:
SGTM